### PR TITLE
Schedules with Parameters

### LINKED ISSUE #3384: Schedules should support parameters
### What is the expected enhancement?
Pulse schedules should support parameters that allow parameterization of the schedule outside of timing. This would allow `schedule` to consume the `Parameter`s of a `QuantumCircuit` and produce a `Schedule` containing these `Parameter`s which could later be fixed.

### PR DESCRIPTION
**Pull request by lcapelluto**
### Summary

ParameterExpressions are supported in Circuits. This PR adds support in Schedules, too. Closes #3384.

### Details and comments

Immediate follow up:
https://github.com/Qiskit/qiskit-terra/compare/master...lcapelluto:add-parameters-attribute?expand=1
  - efficient assignment
  - `parameters` and `is_parameterized` attributes
  - error if assigning parameter which doesn't appear in schedule

Follow up work:
  - drawing
  - ParameterExpressions with complex
  - Support ParameterVector assignment
  - parameterized durations

Context:
 - This is the first step towards deprecating `ParameterizedSchedule`s, which will just become part of `Schedule`.
 - This work is in support of Pulse Gates: circuit parameters can be assigned, and when that happens, parameterized calibrations also need to be assigned. In the future, we should be able to assemble unbound calibrations where the parameters will be bound according to the Gate instruction parameter values.

### CODE REVIEW

File: qiskit/pulse/channels.py [hunk 1]
```diff
@@ -44,8 +45,9 @@ def __init__(self, index: int):
         Raises:
             PulseError: If ``index`` is not a nonnegative integer.
         """
-        if not isinstance(index, int) or index < 0:
-            raise PulseError('Channel index must be a nonnegative integer')
+        if not isinstance(index, ParameterExpression):
+            if not isinstance(index, int) or index < 0:
+                raise PulseError('Channel index must be a nonnegative integer')
```
Comment (taalexander) [inline qiskit/pulse/channels.py:50]:
```suggestion
        if (not isinstance(index, ParameterExpression) 
                and not isinstance(index, int) or index < 0):
            raise PulseError('Channel index must be a nonnegative integer or a ParameterExpression.')
```
https://www.python.org/dev/peps/pep-0008/#multiline-if-statements

File: qiskit/pulse/instructions/acquire.py [hunk 2]
```diff
@@ -17,10 +17,10 @@
 
 from typing import List, Optional
 
-from ..channels import MemorySlot, RegisterSlot, AcquireChannel
-from ..configuration import Kernel, Discriminator
-from ..exceptions import PulseError
-from .instruction import Instruction
+from qiskit.pulse.channels import MemorySlot, RegisterSlot, AcquireChannel
```
Comment (taalexander) [inline qiskit/pulse/instructions/acquire.py:20]: 👍

File: qiskit/pulse/instructions/acquire.py [hunk 3]
```diff
@@ -72,22 +72,19 @@ def __init__(self,
...
+        super().__init__((duration, channel, mem_slot, reg_slot),
                          duration, all_channels, name=name)
 
     @property
     def channel(self) -> AcquireChannel:
         """Return the :py:class:`~qiskit.pulse.channels.Channel` that this instruction is
         scheduled on.
         """
-        return self._channel
+        return self.operands[1]
```
Comment (taalexander) [inline qiskit/pulse/instructions/acquire.py:87]: 👍

File: qiskit/pulse/instructions/instruction.py [hunk 4]
```diff
@@ -219,6 +219,37 @@ def append(self, schedule: ScheduleComponent,
...
+                    new_operands[idx] = new_operands[idx].assign(parameter, value)
+                elif (isinstance(op, Channel)
+                      and isinstance(op.index, ParameterExpression)
+                      and parameter in op.index.parameters):
+                    # TODO: Check int type?
+                    new_index = new_operands[idx].index.assign(parameter, value)
+                    new_operands[idx] = type(op)(new_index)
+
+        if inplace:
+            self._operands = tuple(new_operands)
```
Comment (taalexander) [inline qiskit/pulse/instructions/instruction.py:249]: I'm starting to wonder if we should just switch everything to mutable and request that if they want to do an immutable operation just request they copy and then mutate, eg., `new_inst = copy.copy(inst).assign_parameter(...)`. It might be very cumbersome to always maintain `inplace` going forward. What do you think?

File: qiskit/pulse/instructions/play.py [hunk 5]
```diff
@@ -46,20 +47,34 @@ def __init__(self, pulse: Pulse,
...
-        return self._channel
+        return self.operands[1]
+
+    def assign_parameters(self,
+                          value_dict: Dict[ParameterExpression,
+                                           Union[ParameterExpression, int, float, complex]],
+                          inplace: bool = True) -> 'Play':
+        play_inst = super().assign_parameters(value_dict, inplace)
+
+        if hasattr(self.pulse, 'assign_parameters'):
```
Comment (taalexander) [inline qiskit/pulse/instructions/play.py:72]: Why is this required? Why not a standard method on a pulse and have it do nothing for a waveform?

File: qiskit/pulse/library/parametric_pulses.py [hunk 6]
```diff
@@ -92,6 +93,35 @@ def parameters(self) -> Dict[str, Any]:
...
+        if not any([is_parameterized(val) for val in self.parameters.values()]):
+            return self
+
+        new_parameters = {}
+        for op, op_value in self.parameters.items():
+            for parameter, value in value_dict.items():
+                if is_parameterized(op_value) and parameter in op_value.parameters:
+                    op_value = op_value.assign(parameter, value)
+                    try:
+                        # TODO: some need to be complex, but ParameterExpression doesn't support it
```
Comment (taalexander) [inline qiskit/pulse/library/parametric_pulses.py:118]: This is a bummer. Typing is definitely needed in ParameterExpression.

File: qiskit/pulse/library/parametric_pulses.py [hunk 7]
```diff
@@ -447,3 +486,10 @@ def __init__(self,
         """
         super().__init__(duration, amp, name)
         warnings.warn("The ConstantPulse is deprecated. Use Constant instead", DeprecationWarning)
+
+
+def is_parameterized(value: Any) -> bool:
```
Comment (taalexander) [inline qiskit/pulse/library/parametric_pulses.py:491]: Should this be private to the module?
```suggestion
def _is_parameterized(value: Any) -> bool:
```

**Review by taalexander — COMMENTED** [review]
This looks like a very good! One major question I had is, there a way to get all parameters that may need to be bound?

File: qiskit/pulse/instructions/instruction.py (hunk 4 above)
Comment (lcapelluto) [reply qiskit/pulse/instructions/instruction.py:249]: I definitely agree. I can take that approach here (although, ironically and confusingly, this is one of the only immutable operation in circuits).
Generally, it was very cumbersome to change the default behavior for all operations and the performance was an issue

File: qiskit/pulse/instructions/play.py (hunk 5 above)
Comment (lcapelluto) [reply qiskit/pulse/instructions/play.py:72]: Okay, that makes sense

File: qiskit/pulse/library/parametric_pulses.py (hunk 6 above)
Comment (lcapelluto) [reply qiskit/pulse/library/parametric_pulses.py:118]: I mean ParameterExpressions error on complex values -- I think there might be work in progress to add it

File: qiskit/pulse/library/parametric_pulses.py (hunk 7 above)
Comment (lcapelluto) [reply qiskit/pulse/library/parametric_pulses.py:491]: yes thank you!

File: qiskit/pulse/instructions/instruction.py (hunk 4 above)
Comment (lcapelluto) [reply qiskit/pulse/instructions/instruction.py:249]: @taalexander Do you mean this for schedule and pulse, too?

File: qiskit/pulse/channels.py (hunk 1 above)
Comment (lcapelluto) [reply qiskit/pulse/channels.py:50]: I think it's more readable as is

File: qiskit/pulse/library/parametric_pulses.py (hunk 6 above)
Comment (lcapelluto) [reply qiskit/pulse/library/parametric_pulses.py]: But you are also right about needing typing in ParameterExpression

File: qiskit/circuit/parameterexpression.py [hunk 8]
```diff
@@ -255,6 +272,9 @@ def __int__(self):
                             'cannot be cast to an int.'.format(self.parameters))
         return int(self._symbol_expr)
 
+    def __hash__(self):
+        return hash((tuple(hash(p) for p in self.parameters), str(self)))
```
Comment (kdk) [inline qiskit/circuit/parameterexpression.py:276]:
```suggestion
        return hash((tuple(sorted(self.parameters, key=lambda p: p.name)), str(self)))
```
Iteration order isn't guaranteed for `self.parameters`, which could lead to some hard to track down bugs. ( This is a regular pain point of the parameters API, see https://github.com/Qiskit/qiskit-terra/pull/3910 ).

Comment (kdk) [reply qiskit/circuit/parameterexpression.py:276]: Out of curiosity, where is `ParameterExpression.__hash__` being used? (`ParameterExpression.__eq__` was added for testing the equivalence library, so there might be some edge cases/performance questions on using `srepr` as the key of a large dict.)

Comment (lcapelluto) [reply qiskit/circuit/parameterexpression.py:276]: thank you

File: qiskit/pulse/instructions/instruction.py [hunk 9]
```diff
@@ -221,34 +221,36 @@ def append(self, schedule: ScheduleComponent,
...
         for idx, op in enumerate(self.operands):
             for parameter, value in value_dict.items():
                 if isinstance(op, ParameterExpression) and parameter in op.parameters:
                     new_operands[idx] = new_operands[idx].assign(parameter, value)
                 elif (isinstance(op, Channel)
                       and isinstance(op.index, ParameterExpression)
                       and parameter in op.index.parameters):
-                    # TODO: Check int type?
                     new_index = new_operands[idx].index.assign(parameter, value)
+                    if not new_index.parameters:
```
Comment (taalexander) [inline qiskit/pulse/instructions/instruction.py:245]: The implications of this are interesting. How does this work if there is an are two overlapping instructions with the same parameter, would these not be invalid? Will this be resolved at the binding time?

I am somewhat uncomfortable perhaps because I am unclear with the Qiskit parameter implementation of what is supposed to be macros for code-generation time and what is imagined to be represented as reusable functions/code to be targeted to the backend hardware and executed by the QPU. 

As an analogy to a classical compiler, this is akin to a virtual register, pre-register allocation, but post-scheduling. During compilation, it is fine to re-write registers. However, post-compilation within a compiled program register locations will not be passed as values with resolved scheduling (which is very critical), they will be fixed in the binary and not evaluable. We cannot hope to dynamically at runtime rewrite channels within functions, except by ahead of time monomorphization of these functions.

File: qiskit/pulse/library/waveform.py [hunk 10]
```diff
@@ -93,6 +94,13 @@ def _clip(self, samples: np.ndarray, epsilon: float = 1e-7) -> np.ndarray:
 
         return samples
 
+    def assign_parameters(self,
+                          value_dict: Dict[ParameterExpression,
+                                           Union[ParameterExpression, int, float, complex]]
```
Comment (taalexander) [inline qiskit/pulse/library/waveform.py:99]: I think we should consider defining a type for this such as `PulseValue` or something like that. Type hints are getting long.

File: test/python/pulse/test_parameters.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        """Test Parameter substitution (vs bind)."""
+        schedule = pulse.Schedule()
+        schedule += pulse.SetFrequency(self.alpha, DriveChannel(0))
+
+        schedule.assign_parameters({self.alpha: 2*self.beta})
+        self.assertEqual(schedule.instructions[0][1].frequency, 2*self.beta)
+        schedule.assign_parameters({self.beta: self.freq / 2})
+        self.assertEqual(float(schedule.instructions[0][1].frequency), self.freq)
+
+    def test_channels(self):
```
Comment (taalexander) [inline test/python/pulse/test_parameters.py:111]: Could we please add a test for what happens when there are two overlapping instructions with the same parameter as the channel to define this behaviour?

**Review by taalexander — CHANGES_REQUESTED** [review]

File: qiskit/circuit/parameterexpression.py [hunk 12]
```diff
@@ -255,6 +272,9 @@ def __int__(self):
                             'cannot be cast to an int.'.format(self.parameters))
         return int(self._symbol_expr)
 
+    def __hash__(self):
+        return hash((tuple(sorted(self.parameters, key=lambda p: p.name)), str(self)))
```
Comment (ewinston) [inline qiskit/circuit/parameterexpression.py:276]: Maybe you can avoid the ordering issue with,
`hash((frozenset(self._parameter_symbols), self._symbol_expr))`

File: qiskit/pulse/channels.py [hunk 13]
```diff
@@ -44,8 +45,9 @@ def __init__(self, index: int):
         Raises:
             PulseError: If ``index`` is not a nonnegative integer.
         """
-        if not isinstance(index, int) or index < 0:
-            raise PulseError('Channel index must be a nonnegative integer')
+        if not isinstance(index, ParameterExpression):
```
Comment (ewinston) [inline qiskit/pulse/channels.py:48]: Need to update type hinting to include ParameterExpression?

File: qiskit/pulse/instructions/delay.py [hunk 14]
```diff
@@ -45,12 +45,11 @@ def __init__(self, duration: int,
...
-        self._channel = channel
         super().__init__((duration, channel), duration, (channel,), name=name)
 
     @property
     def channel(self) -> Channel:
         """Return the :py:class:`~qiskit.pulse.channels.Channel` that this instruction is
         scheduled on.
         """
-        return self._channel
+        return self.operands[1]
```
Comment (ewinston) [inline qiskit/pulse/instructions/delay.py:55]: This comment applies to this change to use of `operands` which is a little less clear than the original code. Could this be avoided with something like ParameterTable?

Comment (taalexander) [reply qiskit/pulse/instructions/delay.py:55]: Instruction arguments are an ordered list of operands and need not necessarily be named as their functionality is enforced by the data structure, and is not user-assignable. This is a pretty standard object model in intermediate representations, eg., [see](https://github.com/numba/llvmlite/blob/4acef2d5165768f43b315dd47ed99a5f8f63b3a8/llvmlite/ir/instructions.py#L18). This property just provides an easy way for the user to get a named value.

File: test/python/pulse/test_parameters.py (hunk 11 above)
Comment (lcapelluto) [reply test/python/pulse/test_parameters.py:111]: Instructions have the same parameter as which channel? Do you mean two instructions with parameterized channels? And what would you like the test to be? Verify that assignment of the parameterized channel index on overlapping instructions causes an error?

File: qiskit/pulse/channels.py (hunk 13 above)
Comment (lcapelluto) [reply qiskit/pulse/channels.py:48]: Ooh good point, this applies in many places now

File: qiskit/pulse/instructions/delay.py (hunk 14 above)
Comment (lcapelluto) [reply qiskit/pulse/instructions/delay.py:55]: I think the name of the property conveys the same information. The interface is just as clear, now I'm saying, "the first operand is the channel", rather than "the channel is the channel". I did look into using ParameterTable, but it didn't seem to be a good fit. In particular, here, this is necessary or else `operands[1]` and `_channel` aren't easily both updated upon assignment

File: qiskit/circuit/parameterexpression.py (hunk 12 above)
Comment (lcapelluto) [inline qiskit/circuit/parameterexpression.py:276]:
```suggestion
        return hash((frozenset(self._parameter_symbols), self._symbol_expr))
```

File: qiskit/pulse/library/waveform.py (hunk 10 above)
Comment (lcapelluto) [reply qiskit/pulse/library/waveform.py:99]: I agree. I avoided doing this because it has to be done in every file -- but I think parameterexpression.py is the right place for it

File: qiskit/pulse/instructions/instruction.py (hunk 9 above)
Comment (lcapelluto) [reply qiskit/pulse/instructions/instruction.py:245]: Resolved at assignment time. I'll add a test for it

**Comment by lcapelluto** [comment]
@ewinston @taalexander Very helpful first review! Responded

**Review by taalexander — APPROVED** [review]
